source/common/http/websocket/ws_handler_impl.cc

```diff
@@ -43,22 +46,29 @@ void WsHandlerImpl::onInitFailure(UpstreamFailureReason failure_reason) {
 }
 
 Network::FilterStatus WsHandlerImpl::onData(Buffer::Instance& data, bool end_stream) {
-  ENVOY_LOG(debug, "WsHandlerImpl::onData with buffer length {}, end_stream == {}", data.length(),
+  ENVOY_LOG(trace, "WsHandlerImpl::onData with buffer length {}, end_stream == {}", data.length(),
             end_stream);
 
-  // If we are connected to upstream, then data should have been drained already.
-  // And if we're not connected yet, it is expected that TcpProxy will readDisable(true)
-  // the downstream connection until it is ready to send data to the upstream connection,
-  // so onData() should be called zero or one times before is_connected_ is true.
-  ASSERT(queued_data_.length() == 0);
+  switch (state_) {
+  case ConnectState::PreConnect:
+    // It is expected that TcpProxy will readDisable(true) the downstream connection until it is
+    // ready to send data to the upstream connection, so onData() should be called zero or one times
+    // before state_ becomes Connected.
+    ASSERT(queued_data_.length() == 0);
 
-  if (is_connected_) {
-    ENVOY_LOG(debug, "WsHandlerImpl::onData is connected");
-    return Extensions::NetworkFilters::TcpProxy::TcpProxyFilter::onData(data, end_stream);
-  } else {
-    ENVOY_LOG(debug, "WsHandlerImpl::onData is NOT connected");
+    ENVOY_LOG(trace, "WsHandlerImpl::onData is NOT connected");
     queued_data_.move(data);
     queued_end_stream_ = end_stream;
+    break;
+  case ConnectState::Connected:
+    // Data should have been drained already when we transitioned to Connected.
+    ASSERT(queued_data_.length() == 0);
+
+    ENVOY_LOG(trace, "WsHandlerImpl::onData is connected");
+    return Extensions::NetworkFilters::TcpProxy::TcpProxyFilter::onData(data, end_stream);
+  case ConnectState::Failed:
```

Comment:
Any reason to not readDisable when we send the error response?  This is going to result in dropping the first onData's worth of data then buffering the rest in the network stack until we hit the connection limit.   Seems better to stop reading incoming data even if it should be a fairly fast response as it's headers-only.

Reply:
I felt this made it more robust to future changes. It's already a little weird that TcpProxy is doing readDisable(true) and ConnectionManagerImpl::encodeHeaders() is doing readDisable(false). Adding another actor that modifies the readDisable() state seemed like it would just get more brittle.

Reply:
Fair.  I do think we should readDisable any time we send a response in HTTP/1.1 land but perhaps this is not the place :-)